### ISSUE TITLE
Windows builds produce a large number of compiler and linker warnings

### ISSUE THREAD

**Issue by pauljon** [randombit/botan#2211]
They can be seen with a normal library build, but are more easily seen with single-file amalgamation. Having these warnings complicates the use of the amalgamation in projects where warnings are treated as errors.

BTW, thank you for providing the amalgamation as an option. This coupled with the ability to include only the needed modules is just fantastic.

With fresh Botan Git clone checked out to "2.12.1", these commands yield the warnings.

python configure.py --cc=msvc --os=windows --amalgamation --single-amalgamation-file
nmake

**Comment by randombit** [comment]
I know. :/ I don't use Windows at all, so the only time I see warnings is when I check the CI logs, and sometimes it can be tricky to figure out how to silence them without a compiler in hand to experiment with. But the goal is to have zero warnings, since anything besides that is obnoxious for downstream users.

I'm switching GCC/Clang CI builds to error on all warnings in #2206 and plan to extend that to the Windows builds later, when I have time to grind through all the current warnings that is ...

Even with that, things might no be completely silent when compiling an amalgamation that's included in a larger piece of software since certain MSVC warnings like C4250 are disabled because there is effectively no way (AFAIK) to silence them without changing the design of the software in backwards incompatible ways. You'd still have to silence those when building the amalgamation.

**Comment by pauljon** [comment]
All efforts toward this are appreciated. As a data point, I am also using the amalgamation on a Linux build. There was only a single warning there. So, there is quite a difference between my two data points. A bunch of the warnings were conversions being flagged as well as unreachable code.

**Comment by randombit** [comment]
Can you post the warning you are seeing in the Linux build? Probably will be easy to fix.

**Comment by pauljon** [comment]
Sure, here you go. In generating the source amalgamation, I'm using --minimized-build, so BOTAN_HAS_ASN1 is not defined.

```
<snip>/botan_all.cpp: In function ‘std::vector<unsigned int> Botan::parse_asn1_oid(const string&)’:
<snip>/botan_all.cpp:8812:57: error: unused parameter ‘oid’ [-Werror=unused-parameter]
 std::vector<uint32_t> parse_asn1_oid(const std::string& oid)
                                                         ^~~
```

The matching source

```
std::vector<uint32_t> parse_asn1_oid(const std::string& oid)
   {
#if defined(BOTAN_HAS_ASN1)
   return OID(oid).get_components();
#else
   throw Not_Implemented("ASN1 support not available");
#endif
   }
```

**Comment by randombit** [comment]
Ok good that is already fixed in master by #2163 :)

**Comment by randombit** [comment]
With #2212 all warnings from MSVC 2017/2019 are resolved (modulo four that have to be disabled due to no viable workaround), and in CI we now build with /WX to prevent regressions of this.
